semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -1,270 +1,53 @@
 {-# LANGUAGE RankNTypes #-}
-{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE UndecidableInstances #-}
 module Analysis.Syntax
-( -- * Terms
+( -- * Syntax
   Term(..)
 , subterms
-  -- * Abstract interpretation
-, eval0
-, eval
-  -- * Macro-expressible syntax
-, let'
-, letrec
-  -- * Parsing
-, parseFile
-, parseGraph
-, parseNode
-  -- * Debugging
-, analyzeFile
-, parseToTerm
+, foldTerm
+, paraTerm
+, mendlerTerm
+, mendlerParaTerm
 ) where
 
-import qualified Analysis.Carrier.Statement.State as S
-import           Analysis.Effect.Domain
-import           Analysis.Effect.Env (Env, bind, lookupEnv)
-import           Analysis.Effect.Store
-import           Analysis.File
-import           Analysis.Name (Name, name)
-import           Analysis.Reference as Ref
-import           Control.Applicative (Alternative (..), liftA2, liftA3)
-import           Control.Carrier.Throw.Either (runThrow)
-import           Control.Effect.Labelled
-import           Control.Effect.Reader
-import           Control.Effect.Throw (Throw, throwError)
-import           Control.Exception
-import           Control.Monad (guard)
-import           Control.Monad.IO.Class
-import qualified Data.Aeson as A
-import qualified Data.Aeson.Parser as A
-import qualified Data.Aeson.Types as A
-import qualified Data.ByteString.Lazy as B
-import           Data.Foldable (fold, foldl')
-import           Data.Function (fix)
-import qualified Data.IntMap as IntMap
-import           Data.List (sortOn)
-import           Data.List.NonEmpty (NonEmpty, fromList)
-import           Data.Maybe (listToMaybe)
-import           Data.Monoid (First (..))
 import qualified Data.Set as Set
-import           Data.String (IsString (..))
-import           Data.Text (Text)
-import qualified Data.Vector as V
-import qualified Source.Source as Source
-import           Source.Span
-import           System.FilePath
 
-data Term
-  = Var Name
-  | Noop
-  | Iff Term Term Term
-  | Bool Bool
-  | String Text
-  | Throw Term
-  | Let Name Term Term
-  | Term :>> Term
-  | Import (NonEmpty Text)
-  | Function Name [Name] Term
-  | Call Term [Term]
-  | Locate Span Term
-  deriving (Eq, Ord, Show)
+-- Syntax
 
-infixl 1 :>>
+-- | (Currently) untyped term representations.
+data Term sig v
+  = Var v
+  | Term (sig (Term sig v))
 
-subterms :: Term -> Set.Set Term
-subterms t = Set.singleton t <> case t of
-  Var _          -> mempty
-  Noop           -> mempty
-  Iff c t e      -> subterms c <> subterms t <> subterms e
-  Bool _         -> mempty
-  String _       -> mempty
-  Throw t        -> subterms t
-  Let _ v b      -> subterms v <> subterms b
-  a :>> b        -> subterms a <> subterms b
-  Import _       -> mempty
-  Function _ _ b -> subterms b
-  Call f as      -> subterms f <> foldMap subterms as
-  Locate _ b     -> subterms b
+instance (Eq (sig (Term sig v)), Eq v) => Eq (Term sig v) where
+  Var v1  == Var v2  = v1 == v2
+  Term s1 == Term s2 = s1 == s2
+  _       == _       = False
```

Comment:
The instance head can be cleaned up (and avoid the use of `UndecidableInstances`) using `QuantifiedConstraints`. However:

- for `Ord` you end up needing to supply quantified constraints for both `Eq` and `Ord`, and
- it was causing spurious CI failures for ghc9.2 and I really can't be bothered